### ISSUE TITLE
Test that the compiler/library builds with -Zmir-opt-level=3 -Zvalidate-mir

### ISSUE THREAD

**Issue by saethlin** [rust-lang/rust#105706]
Because currently it doesn't. `-Zmir-opt-level=3` has a handful of problems which are reported elsewhere, but discovering problems with MIR optimizations is often as simple as building the compiler with the optimization enabled and `-Zvalidate-mir`. So we should do that in CI.

Note that all MIR opt levels are supposed to be sound, and known unsoundness is supposed to be gated behind `-Zunsound-mir-opts`, which I am not suggesting we test with.

---
Currently attempting to build the standard library with these flags results in pages of this error:
```
error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:20363 ~ core[b221]::core_simd::masks::mask_impl::{impl#6}::to_bitmask_integer), const_param_did: None }) (after pass DestinationPropagation) at bb2[2]:
                                encountered overlapping memory in `Call` terminator: _0 = <U as ReverseBits>::reverse_bits(move _0, const LANES) -> [return: bb3, unwind: bb6] 
   --> /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/../../portable-simd/crates/core_simd/src/masks/full_masks.rs:217:13
    |    
217 |             bitmask.reverse_bits(LANES)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |    
    = note: delayed at compiler/rustc_const_eval/src/transform/validate.rs:779:26
```
Which looks to me like https://github.com/rust-lang/rust/issues/105428

**Comment by jyn514** [comment]
@saethlin also mentioned to me that some opts are using `-Zmir-opt-level=4` as a minimum threshold, it may be helpful to test that as well (maybe in a separate builder).

`-Zvalidate-mir` does not change the output of the compiler and should be ok to turn on unconditionally for libstd; we might want to be more selective about where we enable it for the compiler itself, though.

**Comment by jyn514** [comment]
Mentoring instructions:
1. Add new `rust.validate_mir_opts = N` option to `Config` and config.toml.example: https://github.com/rust-lang/rust/blob/b0f3940c35d565399dccf8c79f38147b40f2724a/src/bootstrap/config.rs#L49. This option should pass `-Zvalidate-mir -Zmir-opt-level=N` unconditionally for libstd. Test it works locally with `build --stage 1 std`, you should see the ICE @saethlin posted above.
2. Enable that new option on all builders around here: https://github.com/rust-lang/rust/blob/b0f3940c35d565399dccf8c79f38147b40f2724a/src/ci/run.sh#L65. You can test that works by opening a PR (please put `r? @jyn514` in the description) and making sure it shows the same ICE.
3. Assuming that works, we can figure out whether we want to enable this for rustc as well on some builders; maybe we should have separate `validate-mir-opts-std` and `validate-mir-opts-rustc` configs?

**Comment by chenyukang** [comment]
Reproduced it on the PR.
